File: Edu/LinkedList_CycleDetection.js

/**
 * Definition for singly-linked list.
 */
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}


class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {

        const map = new Map();
        let cur = head;
        let index = 0;
        while (cur != null) {

            if (map.has(cur)) {
                return map.get(cur);
            }

            map.set(cur, index);
            index++;
            cur = cur.next;
        }

        return -1;

    }

}

const tail = new ListNode(3);
const head = new ListNode(
    0,
    new ListNode(
        1,
        new ListNode(
            2,
            tail
        )
    )
);
tail.next = head;

const s = new Solution;

console.log(s.hasCycle(head));

File: java/src/main/java/com/leetcode/linkedlist/LinkedListCycleDetection.java
package com.leetcode.linkedlist;

public class LinkedListCycleDetection {
    public static final ListNode TAIL = new ListNode(3);
    public static final ListNode HEAD = new ListNode(0,
            new ListNode(1,
                    new ListNode(2,
                            TAIL)));

    static {
        TAIL.next = HEAD;
    }

    public static void main(String[] args) {
        System.out.println(
                new LinkedListCycleDetection().hasCycle(HEAD)
        );
    }

    public boolean hasCycle(ListNode head) {
        if (head == null)  return false;

        var fast = head;
        var slow = head;

        while (fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) return  true;

        }

        return false;
    }
}
